staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go

```diff
@@ -83,29 +83,81 @@ func (g *Cloud) ToInstanceReferences(zone string, instanceNames []string) (refs
 }
 
 // NodeAddresses is an implementation of Instances.NodeAddresses.
-func (g *Cloud) NodeAddresses(_ context.Context, _ types.NodeName) ([]v1.NodeAddress, error) {
-	internalIP, err := metadata.Get("instance/network-interfaces/0/ip")
-	if err != nil {
-		return nil, fmt.Errorf("couldn't get internal IP: %v", err)
+func (g *Cloud) NodeAddresses(ctx context.Context, nodeName types.NodeName) ([]v1.NodeAddress, error) {
+	timeoutCtx, cancel := context.WithTimeout(ctx, 1*time.Hour)
+	defer cancel()
+
+	instanceName := string(nodeName)
+
+	if g.useMetadataServer {
+		// Use metadata server if possible
+		if g.isCurrentInstance(instanceName) {
+
+			nics, err := metadata.Get("instance/network-interfaces/")
+			if err != nil {
+				return nil, fmt.Errorf("couldn't get network interfaces: %v", err)
+			}
+
+			nicsArr := strings.Split(nics, "/\n")
+			nodeAddresses := []v1.NodeAddress{}
+
+			for _, nic := range nicsArr {
+
+				if nic == "" {
+					continue
+				}
+
+				internalIP, err := metadata.Get(fmt.Sprintf("instance/network-interfaces/%v/ip", nic))
+				if err != nil {
+					return nil, fmt.Errorf("couldn't get internal IP: %v", err)
+				}
+				nodeAddresses = append(nodeAddresses, v1.NodeAddress{Type: v1.NodeInternalIP, Address: internalIP})
+
+				acs, err := metadata.Get(fmt.Sprintf("instance/network-interfaces/%v/access-configs/", nic))
+				if err != nil {
+					return nil, fmt.Errorf("couldn't get access configs: %v", err)
+				}
+
+				acsArr := strings.Split(acs, "/\n")
+
+				for _, ac := range acsArr {
+
+					if ac == "" {
+						continue
+					}
+
+					externalIP, err := metadata.Get(fmt.Sprintf("instance/network-interfaces/%v/access-configs/%v/external-ip", nic, ac))
+					if err != nil {
+						return nil, fmt.Errorf("couldn't get external IP: %v", err)
```

Comment:
Discussion: should we print a warning and skip, or should we fail at once?

Reply:
The GET won't fail if there is no external IP, only if the metadata server doesn't respond for some reason